### PR TITLE
logical_and, logical_or

### PR DESCRIPTION
**Pull request by Oceania2018**

### CODE REVIEW

File: src/NumSharp.Core/Logic/np.logical_and.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        /// <summary>
+        ///     Compute the truth value of x1 AND x2 element-wise.
+        /// </summary>
+        /// <param name="lhs">Input boolean array.</param>
+        /// <param name="rhs">Input boolean array.</param>
+        /// <returns>Returns True if the arrays are equal.</returns>
+        /// <remarks>https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_and.html</remarks>
+        public static unsafe NDArray<bool> logical_and(NDArray lhs, NDArray rhs)
+        {
+            if (lhs.typecode == NPTypeCode.Boolean && rhs.typecode == NPTypeCode.Boolean)
```
Comment (Nucs) [inline src/NumSharp.Core/Logic/np.logical_and.cs]: Logical And can happen with any types. not just boolean vs boolean. please handle the rest of the cases. I would suggest using Regen.

File: src/NumSharp.Core/Logic/np.logical_or.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        /// <summary>
+        ///     Compute the truth value of x1 OR x2 element-wise.
+        /// </summary>
+        /// <param name="lhs">Input boolean array.</param>
+        /// <param name="rhs">Input boolean array.</param>
+        /// <returns>Returns True if the arrays are equal.</returns>
+        /// <remarks>https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_or.html</remarks>
+        public static unsafe NDArray<bool> logical_or(NDArray lhs, NDArray rhs)
+        {
+            if (lhs.typecode == NPTypeCode.Boolean && rhs.typecode == NPTypeCode.Boolean)
```
Comment (Nucs) [inline src/NumSharp.Core/Logic/np.logical_or.cs]: Same as above.

**Review by Nucs — CHANGES_REQUESTED** [review]

File: src/NumSharp.Core/Logic/np.logical_and.cs (hunk 1 above)
Comment (Oceania2018) [reply src/NumSharp.Core/Logic/np.logical_and.cs]: I'd say we only handl bool is good.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_or.html

image

Comment (Nucs) [reply src/NumSharp.Core/Logic/np.logical_and.cs]:
```python
import numpy as np
>>> np.logical_and(np.arange(10), np.arange(10))
[False  True  True  True  True  True  True  True  True  True]
```
Behind the scenes the np.arange(10) is casted to ndarray of bool by `x != 0`.
This should be done to all dtypes.

File: src/NumSharp.Core/Logic/np.logical_and.cs [hunk 3]
```diff
@@ -16,7 +16,70 @@ public static partial class np
         /// <remarks>https://docs.scipy.org/doc/numpy/reference/generated/numpy.logical_and.html</remarks>
         public static unsafe NDArray<bool> logical_and(NDArray lhs, NDArray rhs)
         {
-            if (lhs.typecode == NPTypeCode.Boolean && rhs.typecode == NPTypeCode.Boolean)
+#if _REGEN
+            if(lhs.typecode != rhs.typecode)
+            {
+                throw new NotImplementedException("please make sure operands have the same data type");
```
Comment (Nucs) [inline src/NumSharp.Core/Logic/np.logical_and.cs]: They do not have to be of the same dtype.
Like all operators, this has to support all dtypes combinations.

**Review by Nucs — CHANGES_REQUESTED** [review]

**Pull request closed without merging**